include/nektar_interface/function_basis_evaluation.hpp

```diff
@@ -167,6 +168,111 @@ class FunctionEvaluateBasis : public BasisEvaluateBase<T> {
+      ParticleSubGroupSharedPtr particle_sub_group, Sym<COMPONENT_TYPE> sym,
+      const int component) {
+
+    auto particle_group = particle_sub_group->get_particle_group();
+    if (particle_sub_group->is_entire_particle_group()) {
+      return this->evaluate_inner(evaluation_type, particle_group, sym,
+                                  component);
+    }
+
+    const ShapeType shape_type = evaluation_type.get_shape_type();
+    const int cells_iterset_size = this->map_shape_to_count.at(shape_type);
+    if (cells_iterset_size == 0) {
+      return sycl::event{};
+    }
+    const auto h_cells_iterset =
+        this->map_shape_to_dh_cells.at(shape_type)->h_buffer.ptr;
+
+    const auto k_global_coeffs = this->dh_global_coeffs.d_buffer.ptr;
+    const auto k_coeffs_offsets = this->dh_coeffs_offsets.d_buffer.ptr;
+    const auto k_nummodes = this->dh_nummodes.d_buffer.ptr;
+
+    // jacobi coefficients
+    const auto k_coeffs_pnm10 = this->dh_coeffs_pnm10.d_buffer.ptr;
+    const auto k_coeffs_pnm11 = this->dh_coeffs_pnm11.d_buffer.ptr;
+    const auto k_coeffs_pnm2 = this->dh_coeffs_pnm2.d_buffer.ptr;
+    const int k_stride_n = this->stride_n;
+
+    // Local space required
+    const int k_max_total_nummodes0 =
+        this->map_total_nummodes.at(shape_type).at(0);
+    const int k_max_total_nummodes1 =
+        this->map_total_nummodes.at(shape_type).at(1);
+    const int k_max_total_nummodes2 =
+        this->map_total_nummodes.at(shape_type).at(2);
+    const std::size_t local_size_bytes =
+        static_cast<size_t>(k_max_total_nummodes0 + k_max_total_nummodes1 +
+                            k_max_total_nummodes2) *
+        sizeof(REAL);
+    auto local_space = std::make_shared<LocalMemoryBlock>(local_size_bytes);
+
+    const int k_component = component;
+    const int k_ndim = evaluation_type.get_ndim();
+
+    for (std::size_t cx = 0; cx < cells_iterset_size; cx++) {
+      const int cellx = h_cells_iterset[cx];
+      particle_loop(
+          "FunctionEvaluateBasis::ParticleSubGroup", particle_sub_group,
+          [=](auto LOCAL_SPACE, auto REF_POSITIONS, auto OUTPUT) {
+            const REAL *dofs = &k_global_coeffs[k_coeffs_offsets[cellx]];
+            ExpansionLooping::JacobiExpansionLoopingInterface<EVALUATE_TYPE>
+                loop_type{};
+
+            // Get the number of modes in x,y and z.
+            const int nummodes = k_nummodes[cellx];
+
+            REAL xi0, xi1, xi2, eta0, eta1, eta2;
+            xi0 = REF_POSITIONS.at(0);
+            if (k_ndim > 1) {
+              xi1 = REF_POSITIONS.at(1);
+            }
+            if (k_ndim > 2) {
+              xi2 = REF_POSITIONS.at(2);
+            }
+
+            loop_type.loc_coord_to_loc_collapsed(xi0, xi1, xi2, &eta0, &eta1,
+                                                 &eta2);
+
+            // Get the local space for the 1D evaluations in each dimension.
+            REAL *local_space_0 = static_cast<REAL *>(LOCAL_SPACE.data());
```

Comment:
I'm not sure I follow, which lines are you suggesting I wrap in a function (bear in mind this code is hopefully being replaced soon)?

Reply:
I was talking about lines 246-248, as they appeared again in function_basis_projection.hpp. I am not sure if it will be needed elsewhere, and as you say this will be replaced soon, I leave it to your judgement.

Reply:
Re-assessed: probably worth doing. Moved common implementation into structs/free functions for the kernel to call.